File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Views/Slider.kt
package com.abracode.actionui.Views

import androidx.compose.material3.Slider
import androidx.compose.material3.SliderDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.abracode.actionui.Common.ActionUIElement
import com.abracode.actionui.Common.ActionUILogger
import com.abracode.actionui.Common.ActionUIModel
import com.abracode.actionui.Common.ActionUIValueType
import com.abracode.actionui.Common.ActionUIViewConstruction
import com.abracode.actionui.Common.LocalActionUILogger
import com.abracode.actionui.Common.LocalWindowModel
import com.abracode.actionui.Common.LoggerLevel
import com.abracode.actionui.Helpers.LocalActionUIEnabled
import com.abracode.actionui.Helpers.LocalActionUITint
import com.abracode.actionui.Helpers.numberProperty
import com.abracode.actionui.Helpers.stringProperty
import kotlinx.serialization.json.JsonObject
import kotlin.math.roundToInt

/**
 * Continuous (or stepped) value selector along a linear range. Mirror of the
 * Apple `Slider` element (`ActionUI/Views/Slider.swift`), which wraps
 * `SwiftUI.Slider`. The first **Double**-valued control on the Android state
 * bridge (entry 17): it declares [ActionUIValueType.DOUBLE], so a host reads its
 * position with `ActionUIModel.getElementValue(...)` and sets it with
 * `setElementValueFromString(..., "75")`, and the control recomposes.
 *
 * Rendered as a Material3 [Slider]. Property mapping:
 *
 *   * `value` - initial position (Double, defaults to `0.0`), seeded into the
 *     element's [com.abracode.actionui.Common.ViewModel] at populate time.
 *   * `range` - `{ "min": .., "max": .. }` (defaults to `0.0..1.0`; `min <= max`).
 *   * `step`  - increment; absent or non-positive => continuous. Clamped so it
 *     never exceeds the range size, matching the Apple validator.
 *   * `valueChangeActionID` - dispatched through [ActionUIModel] on every change,
 *     the analog of the Apple binding's `set` closure.
 *
 * **Float backing (the one platform divergence).** SwiftUI's `Slider` is
 * `Double`; Compose's is `Float`. The value is stored as `Double` (so the value
 * API stays `Double`-typed) and converted to `Float` for the control and back on
 * change. A continuous drag can therefore land on a value with `Float` precision
 * artifacts; stepped sliders stay clean. This is inherent to the Compose control.
 *
 * **Compose `steps` != step size.** Compose's `steps` is the number of discrete
 * stops *between* the endpoints, not the increment. [sliderStepCount] converts
 * a step size into that count.
 *
 * **State.** ViewModel-backed when a [com.abracode.actionui.Common.WindowModel]
 * is in scope (host binding; needs a positive `id`), else local
 * [rememberSaveable] - the same dual-path binding the other controls use.
 */
object Slider : ActionUIViewConstruction {
    override val valueType = ActionUIValueType.DOUBLE

    override fun initialValue(element: ActionUIElement): Any? =
        element.properties?.numberProperty("value") ?: 0.0

    @Composable
    override fun BuildView(element: ActionUIElement, modifier: Modifier) {
        val props = element.properties
        val logger = LocalActionUILogger.current

        val config = resolveSliderConfig(props, logger)
        val valueChangeActionID = props?.stringProperty("valueChangeActionID")
        val initial = props?.numberProperty("value") ?: 0.0

        // Bind to the ViewModel value when a window is in scope; else local state.
        val viewModel = LocalWindowModel.current?.viewModels?.get(element.id)
        var localValue by rememberSaveable(element.id) { mutableStateOf(initial) }
        val current = if (viewModel != null) (viewModel.value as? Double) ?: initial else localValue

        val stepCount = sliderStepCount(config.min, config.max, config.step)

        // SwiftUI `.tint` colors the slider's thumb and the filled (active) part
        // of the track; an inherited tint maps to both.
        val tint = LocalActionUITint.current
        val colors = if (tint != null) {
            SliderDefaults.colors(thumbColor = tint, activeTrackColor = tint)
        } else {
            SliderDefaults.colors()
        }

        Slider(
            value = current.coerceIn(config.min, config.max).toFloat(),
            onValueChange = { new ->
                val asDouble = new.toDouble()
                if (asDouble != current) {
                    if (viewModel != null) viewModel.value = asDouble else localValue = asDouble
                    if (valueChangeActionID != null) {
                        ActionUIModel.actionHandler(valueChangeActionID, viewID = element.id, viewPartID = 0)
                    }
                }
            },
            modifier = modifier,
            // SwiftUI `.disabled` (set here or on any ancestor) -> M3 `enabled`.
            enabled = LocalActionUIEnabled.current,
            valueRange = config.min.toFloat()..config.max.toFloat(),
            steps = stepCount,
            colors = colors,
        )
    }
}

/** Resolved, validated slider bounds and (optional) step size. */
internal data class SliderConfig(val min: Double, val max: Double, val step: Double?)

/**
 * Resolves and validates `range`/`step` into a [SliderConfig], mirroring the
 * Apple `Slider.validateProperties`: an invalid/missing `range` defaults to
 * `0.0..1.0`; a non-positive or non-numeric `step` defaults to `1.0`; and a
 * step larger than the range size is clamped to it. Pure (logging aside) so it
 * is unit-testable.
 */
internal fun resolveSliderConfig(props: JsonObject?, logger: ActionUILogger): SliderConfig {
    var min = 0.0
    var max = 1.0
    val rangeObj = props?.get("range") as? JsonObject
    if (rangeObj != null) {
        val rMin = rangeObj.numberProperty("min")
        val rMax = rangeObj.numberProperty("max")
        if (rMin != null && rMax != null && rMin <= rMax) {
            min = rMin
            max = rMax
        } else {
            logger.log(
                "Slider range must have valid min/max numbers with min <= max; defaulting to 0.0..1.0",
                LoggerLevel.warning
            )
        }
    }

    var step: Double? = null
    if (props?.get("step") != null) {
        val rangeSize = max - min
        val rawStep = props.numberProperty("step")
        step = if (rawStep != null && rawStep > 0.0) {
            if (rawStep > rangeSize) {
                logger.log(
                    "Slider step must not exceed range (max - min); clamping to $rangeSize",
                    LoggerLevel.warning
                )
                rangeSize
            } else {
                rawStep
            }
        } else {
            logger.log("Slider step must be a positive number; defaulting to 1.0", LoggerLevel.warning)
            if (1.0 > rangeSize) rangeSize else 1.0
        }
    }
    return SliderConfig(min, max, step)
}

/**
 * Converts a step *size* into the Compose `steps` value - the number of discrete
 * stops between the endpoints (`(max - min) / step - 1`). Returns `0` (continuous)
 * when there is no positive step or the range is degenerate.
 */
internal fun sliderStepCount(min: Double, max: Double, step: Double?): Int {
    if (step == null || step <= 0.0 || max <= min) return 0
    return (((max - min) / step).roundToInt() - 1).coerceAtLeast(0)
}

File: ActionUIWeb/src/Views/Slider.js
// Slider.js — Slider element.
// Web analog of ActionUI/Views/Slider.swift (and ActionUIAndroid Views/Slider.kt).
//
// A continuous (or stepped) value selector along a linear range, rendered as an
// <input type="range">. The first Double-valued control on the web state bridge:
// valueType is "double", so a host reads the position with get/setDouble(id) and
// the thumb tracks it.
//
// Properties (mirroring Slider.swift):
//   value  initial position (Double, defaults to 0.0).
//   range  { min, max } (defaults to 0.0…1.0; min <= max).
//   step   increment; absent or non-positive ⇒ continuous (input step="any").
//          Clamped so it never exceeds the range size, matching the validator.
//   valueChangeActionID  dispatched on every change (the analog of the Apple
//          binding's `set` closure), including programmatic setElementValue.

import { register } from "../Common/ActionUIRegistry.js";
import { markHandlesAction } from "../Common/ModifierResolver.js";

const DEFAULT_RANGE = { min: 0.0, max: 1.0 };

register("Slider", {
    valueType: "double",

    // Mirrors Slider.swift validateProperties (warning text included verbatim).
    validateProperties: (properties, logger) => {
        const validated = { ...properties };

        // value: a non-number (but present) defaults to 0.0; absent ⇒ 0.0.
        if (typeof validated.value !== "number") {
            if (validated.value !== undefined) {
                logger.log("Slider value must be a number; defaulting to 0.0", "warning");
            }
            validated.value = 0.0;
        }

        // range: a dictionary with numeric min/max and min <= max, else 0.0…1.0.
        if (isPlainObject(validated.range)) {
            const { min, max } = validated.range;
            if (typeof min === "number" && typeof max === "number" && min <= max) {
                validated.range = { min, max };
            } else {
                logger.log("Slider range must have valid min/max numbers with min <= max; defaulting to 0.0...1.0", "warning");
                validated.range = { ...DEFAULT_RANGE };
            }
        } else if (validated.range !== undefined) {
            logger.log("Slider range must be a dictionary with min/max numbers; defaulting to 0.0...1.0", "warning");
            validated.range = { ...DEFAULT_RANGE };
        }

        // step: a positive number, clamped to the range size; a valid range is
        // required for a step (matching "range becomes required if you specify
        // step"), else it defaults to 1.0.
        if (typeof validated.step === "number" && validated.step > 0) {
            if (isPlainObject(validated.range)
                && typeof validated.range.min === "number" && typeof validated.range.max === "number") {
                const rangeSize = validated.range.max - validated.range.min;
                if (validated.step > rangeSize) {
                    logger.log(`Slider step must not exceed range (max - min); clamping to ${formatDouble(rangeSize)}`, "warning");
                    validated.step = rangeSize;
                }
            } else {
                logger.log("Slider step requires a valid range; defaulting to 1.0", "warning");
                validated.step = 1.0;
            }
        } else if (validated.step !== undefined) {
            logger.log("Slider step must be a positive number; defaulting to 1.0", "warning");
            validated.step = 1.0;
        }

        return validated;
    },

    initialValue: (element, properties) => properties.value ?? 0.0,

    buildView: (element, properties, ctx) => {
        const min = properties.range?.min ?? DEFAULT_RANGE.min;
        const max = properties.range?.max ?? DEFAULT_RANGE.max;
        const initial = properties.value ?? 0.0;

        const input = document.createElement("input");
        input.type = "range";
        input.className = "aui-slider";
        input.min = String(min);
        input.max = String(max);
        // A positive step snaps; otherwise "any" keeps the slide continuous (the
        // browser default is 1, which would quantize a 0…1 range to its endpoints).
        input.step = (typeof properties.step === "number" && properties.step > 0)
            ? String(properties.step) : "any";
        input.value = String(clamp(initial, min, max));

        const dispatchValueChange = () => {
            if (typeof properties.valueChangeActionID === "string") {
                ctx.model.dispatchAction(properties.valueChangeActionID, element.id);
            }
        };

        markHandlesAction(input);
        // `input` fires on every position change during a drag (SwiftUI's binding
        // set); `change` would only fire on release.
        input.addEventListener("input", dispatchValueChange);

        if (element.id > 0) {
            ctx.model.bind(element.id, {
                getValue: () => Number(input.value),
                setValue: (value) => {
                    const n = Number(value);
                    input.value = String(clamp(Number.isFinite(n) ? n : initial, min, max));
                    dispatchValueChange();
                },
            });
        }
        return input;
    },
});

function isPlainObject(value) {
    return value !== null && typeof value === "object" && !Array.isArray(value);
}

// Formats a number the way Swift/Kotlin interpolate a Double, so the warning text
// matches the reference platforms (an integer-valued Double prints "10.0", not
// "10"). Non-integers use JS's shortest round-trip form, as the others do.
function formatDouble(n) {
    return Number.isInteger(n) ? n.toFixed(1) : String(n);
}

function clamp(value, min, max) {
    return Math.min(Math.max(value, min), max);
}
